R/CohortSample.R

```diff
@@ -0,0 +1,285 @@
+#' @noRd
+.sampleCohort <- function(connection,
+                          targetCohortId,
+                          targetTable,
+                          outputCohortId,
+                          outputTable,
+                          cohortDatabaseSchema,
+                          outputDatabaseSchema,
+                          sampleTable,
+                          seed,
+                          tempEmulationSchema) {
+
+  randSampleTableName <- paste0("#SAMPLE_TABLE_", seed)
+  DatabaseConnector::insertTable(connection = connection,
+                                 data = sampleTable,
+                                 dropTableIfExists = TRUE,
+                                 tempTable = TRUE,
+                                 tempEmulationSchema = tempEmulationSchema,
+                                 tableName = randSampleTableName)
+
+  execSql <- SqlRender::readSql(system.file("sql", "sql_server", "sampling", "RandomSample.sql", package = "CohortGenerator"))
+  DatabaseConnector::renderTranslateExecuteSql(connection,
+                                               execSql,
+                                               tempEmulationSchema = tempEmulationSchema,
+                                               random_sample_table = randSampleTableName,
+                                               target_cohort_id = targetCohortId,
+                                               output_cohort_id = outputCohortId,
+                                               cohort_database_schema = cohortDatabaseSchema,
+                                               output_database_schema = outputDatabaseSchema,
+                                               output_table = outputTable,
+                                               target_table = targetTable)
+}
+
+
+.computeIdentifierExpression <- function(identifierExpression, cohortId, seed) {
+  allowed_vars <- c("cohortId", "seed")
+  vars_in_expression <- intersect(all.vars(parse(text = identifierExpression)), allowed_vars)
+
+  if (length(setdiff(all.vars(parse(text = identifierExpression)), vars_in_expression)) > 0) {
+    stop("Invalid variable in expression.")
+  }
+
+  expr <- parse(text = identifierExpression)
+  result <- eval(expr, list(cohortId = cohortId, seed = seed))
+  return(result)
+}
+
+
+.checkUniqueOutputIds <- function(cohortIds, seed, identifierExpression, cohortTableNames) {
+  idSet <- purrr::map2(cohortIds, seed, function(.x, .s) {
+    .computeIdentifierExpression(identifierExpression, .x, .s)
+  })
+
+  # If output is a different table to base table then target ids don't need to be distinct from output
+  if (cohortTableNames$cohortTable == cohortTableNames$cohortSampleTable) {
+    idSet <- c(idSet, cohortIds)
+  }
+  errorMessage <- "identifier expression does not produce unique output for cohort ids"
+  if(length(unique(idSet)) != length(idSet)) stop(errorMessage)
+  invisible(NULL)
+}
+
+#' Sample Cohort Definition Set
```

Comment:
Let's make it clear that these sampling methods are for software development purposes only and should not be used for real observational studies.